### ISSUE TITLE
Circular inclusions can lose data

### ISSUE THREAD

**Issue by drakedowsett** [JSONAPI-Resources/jsonapi-resources#802]
Verified on `0.8.0.beta2` in a simple testing app with three models: `Box`, `Thing`, `User`, where:

```
boxes ------> things
       many    | ^
           one | |
               | | many
               v |
              users
```

When requesting

```
/api/boxes?include=things,things.user
```

everything looks ok, but when requesting one level deeper for

```
/api/boxes?include=things,things.user,things.user.things
```

so that the `User` includes its `Thing`s, some of the thing-to-user relationship data disappears.

It seems that there's a depth-first search that explores the first thing-to-user relationship via the first `Thing` it encounters and finds the remaining `Thing`s  through the user-to-things relationship without including the thing-to-user relationship. Requesting

```
/api/boxes?include=things,things.user,things.user.things,things.user.things.user
```

with yet another circular reference resolves this, corroborating the depth-first hypothesis.

An example: one `Box` with two `Thing`s, each with the same `User`. Requests and results of `included` list (with `links` and `box` relationship removed):

```
# /api/boxes?include=things,things.user
...
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {
          "data": {
            "type": "users", "id": "1"
          }
        }
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {
          "data": {
            "type": "users", "id": "1"
          }
        }
      }
    },
    {
      "id": "1",
      "type": "users",
      "relationships": {
        "things": {}
      }
    }
...
```

```
...
# /api/boxes?include=things,things.user,things.user.things
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {}
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {
          "data": {
            "type": "users", "id": "1"
          }
        }
      }
    },
    {
      "id": "1",
      "type": "users",
      "relationships": {
        "things": {
          "data": [
            {
              "type": "things", "id": "1"
            },
            {
              "type": "things", "id": "2"
            }
          ]
        }
      }
    }
...
```

...`Thing` loses its `User` data. To get both the things-to-user data and also the user-to-things data, I am having to request:

```
# /api/boxes?include=things,things.user,things.user.things,things.user.things.user
...
    {
      "id": "1",
      "type": "things",
      "relationships": {
        "user": {
          "data": {
            "type": "users", "id": "1"
          }
        }
      }
    },
    {
      "id": "2",
      "type": "things",
      "relationships": {
        "user": {
          "data": {
            "type": "users", "id": "1"
          }
        }
      }
    },
    {
      "id": "1",
      "type": "users",
      "relationships": {
        "things": {
          "data": [
            {
              "type": "things", "id": "1"
            },
            {
              "type": "things", "id": "2"
            }
          ]
        }
      }
    }
...
```

My testing app setup:

```
Rails.application.routes.draw do
  namespace :api do
    jsonapi_resources :boxes
  end
end

class Box < ActiveRecord::Base
  has_many :things
end

class Thing < ActiveRecord::Base
  belongs_to :box
  belongs_to :user
end

class User < ActiveRecord::Base
  has_many :things
end

class Api::BoxesController < ApplicationController
  include JSONAPI::ActsAsResourceController
end

class Api::BoxResource < JSONAPI::Resource
  has_many :things
end

class Api::ThingResource < JSONAPI::Resource
  has_one :box
  has_one :user
end

class Api::UserResource < JSONAPI::Resource
  has_many :things
end
```

**Comment by drakedowsett** [comment]
Also verified on `master` (at `cdb933dc57709694c8dea7d364586872eec7454d`).

**Comment by drakedowsett** [comment]
I have found in the v0.8.0.beta2 tag that changing this `merge!` to a `deep_merge!` resolves the issue: https://github.com/cerebris/jsonapi-resources/blob/v0.8.0.beta2/lib/jsonapi/resource_serializer.rb#L351

Have not looked at latest master, but imagine there is a similar issue.

/cc @DavidMikeSimon

**Comment by lgebhardt** [comment]
@drakedowsett On master I can see the test fail using your test app. However I'm not getting the error to show using the beta branch at all. So I can't confirm if the `deep_merge!` is fixing things. Can you check again?

**Comment by lgebhardt** [comment]
@drakedowsett Could you take a look at #831 and let me know if this fixes the issue. I added tests based on your test project. One change was I added reciprocal `related_things`. The change to `deep_merge` was needed to fix the issue. Thanks for your work on this.

**Comment by drakedowsett** [comment]
@lgebhardt @DavidMikeSimon I can confirm this is addressed in v0.8.0 and master (e53a32daa4bd)! Thanks so much! Will check #803 now too.
